src/DataStructures/Tensor/Expressions/TensorExpression.hpp

```diff
@@ -63,14 +63,29 @@ struct TensorIndex {
 
 /*!
  * \ingroup TensorExpressionsGroup
- * Given a TensorIndex type T, returns the TensorIndex type with opposite
- * valence. For example, if `T == ti_a_t`, then
- * `tensorindex_with_opposite_valence == ti_A_t`
+ * \brief Returns the TensorIndex value of with opposite valence.
+ *
+ * \details The input value represents a TensorIndex value, which encodes
+ * both the valence of the index and whether the index is spacetime or
+ * spatial. This function returns the value that corresponds to the encoding of
+ * the TensorIndex with the same index type, but opposite valence.
+ *
+ * For example, 0 is the TensorIndex value for `ti_a_t`. If `i == 0`,
+ * then 100 will be returned, which is the TensorIndex value for `ti_A_t`.
+ * If `i == 100` (representing `ti_A_t`), then 0 (representing `ti_a_t`) is
+ * returned.
+ *
+ * @param i a TensorIndex value that represents a generic index
+ * @return the TensorIndex value that encodes the generic index with the
+ * opposite valence
  */
-template <typename T>
-using tensorindex_with_opposite_valence =
-    TensorIndex<T::valence == UpLo::Up ? T::value - upper_sentinel
-                                       : T::value + upper_sentinel>;
+SPECTRE_ALWAYS_INLINE static constexpr size_t
+get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
+  return ((i >= upper_sentinel and i < spatial_sentinel) or
```

Comment:
[optional] you no longer need this to be a ternary. and if-else block might be easier to read

Reply:
Interestingly, when this is converted into `if constexpr` / `else`, the compiler complains about these expressions not being evaluable as `constexpr`:

```
SPECTRE_ALWAYS_INLINE static constexpr size_t
get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
  if constexpr ((i >= upper_sentinel and i < spatial_sentinel) or
                (i >= upper_spatial_sentinel)) {
    // `i` represents an upper index, so return the lower index's encoding
    return i - upper_sentinel;
  } else {
    // `i` represents a lower index, so return the upper index's encoding
    return i + upper_sentinel;
  }
}
```

```
SPECTRE_ALWAYS_INLINE static constexpr size_t
get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
  constexpr bool is_upper = ((i >= upper_sentinel and i < spatial_sentinel) or
                             (i >= upper_spatial_sentinel));
  if constexpr (is_upper) {
    // `i` represents an upper index, so return the lower index's encoding
    return i - upper_sentinel;
  } else {
    // `i` represents a lower index, so return the upper index's encoding
    return i + upper_sentinel;
  }
}
```

It looks to me like if I want to turn this ternary into an `if constexpr`/`else`, then the function will need to be templated on the value instead of it being passed in as a normal argument. The `if constexpr`/`else` might be somewhat clearer, but then I also don't need to know if we need to create so many instantiations of the function template if it works as is.

Let me know your thoughts. Maybe I'm misunderstanding what you were suggesting, here!

Reply:
It is now an if-else!